helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -146,6 +147,11 @@ private double getAssignmentNormalizedScore(AssignableNode node, AssignableRepli
     Map<String, ResourceAssignment> baselineAssignment =
         clusterModel.getContext().getBaselineAssignment();
 
+    Map<String, Integer> replicaHashCodeMap = orderedAssignableReplicas.parallelStream().collect(
+        Collectors.toMap(AssignableReplica::toString,
+            replica -> Objects.hash(replica.toString(), clusterModel.getAssignableNodes().keySet()),
```

Comment:
It would be simpler if you add a `hashcode()` method inside the AssignableReplica class.
It becomes 
```
Map<String, Integer> replicaHashCodeMap = orderedAssignableReplicas.parallelStream().collect(
        Collectors.toMap(AssignableReplica::toString, AssignableReplica::hashCode);
```

Reply:
And do you really need this level of parallism? If so, you might want to update stream() in line 142 as well

Reply:
That will break some logic, there is no real hashcode logic for replica. Here the code just serves for ordering. If I add it to the hashcode() it will impact the maps that use AssignableRepilca as the key.

Regarding 142, since it is not computing-intensive, it seems not improved a lot when I tried. So I just don't do it.